content/posts/2024-12-04-improve-clang-doc.md

```diff
@@ -0,0 +1,114 @@
+---
+author: "Peter Chou"
+date: "2024-12-04"
+tags: ["GSoC", "clang-doc"]
+title: "GSoC 2024: Improve Clang Doc"
+---
+
+Hi, my name is Peter, and this year I was involved in Google Summer of Code 2024. I worked on [improving the clang-doc documenation generator](https://discourse.llvm.org/t/improve-clang-doc-usability/76996)
+
+Mentors: Petr Hosek and Paul Kirth
+
+## Project Background
+
+Clang-Doc is a documentation generator developed on top of libtooling, as an
+alternative to Doxygen. Development started in 2018, and continued through 2019,
+however it has since stalled. Currently, the tool can generate HTML, YAML, and markdown but the generated output has usability issues. This GSOC project was aimed to address the pain points regarding the output of the HTML, by adding support for various C++ constructs and reworking the css of the HTML output to be more user friendly.
+
+## Work Done
+
+The original scope of the project was to improve the output of clang-doc's generation. However during testing the tool was significantly slower than expected which made developing features for the tool impossible. 
+Each compilation of the LLVM codebase was taking upwards of 10 hours on my local machine. Additionally the tool utilized a lot of memory and was prone to crashing with an out of memory error. Similar tools such as Doxygen and Hdoc ran in comparatively less time for the same codebase. This pointed to a significant bottleneck within clang-doc’s codepath when generating large scale software projects. Due to this the project scope quickly changed to improving the runtime of clang-doc so that it could run much faster. It was only during the latter half of the project did the scope change back to improving clang-doc’s generation.
+
+### Added More Test Cases to Clang-Doc test suite
+
+
+Clang-doc previously had test’s which did not test the full scope of the tool especially with regards to the HTML or Markdown output of the tool. In order to make sure the optimization experiments were not accidentally breaking the tool, it was necessary to add more end to end tests. 
+In summary I added four comprehensive tests which covered all features that we were not testing such as testing the generation for Enums, Namespace and Records for HTML and Markdown. 
+
+
+### Improve clang-doc’s performance by 1.58 times 
+
+Internally the way clang-doc works is by leveraging libtooling's ASTVisitor class to parse the declaration in each project and serializing it into an internal format, which gets deserialized later when we output the final format. The bottleneck in clang-doc lied in the way clang-doc was doing redundant work when it was visiting each declaration. 
```

Comment:
```suggestion
Internally, the way Clang-Doc works is by leveraging libtooling's ASTVisitor class to parse the source level declarations in each TU and serializing it into an internal format which gets deserialized later when we output the final format. Eventually, we were able to identify a major bottleneck in clang-doc's performance to doing redundant work when it was processing each declaration. We settled on a caching/memoization strategy to minimize the redundant work.
```

Whatever capitalization you decide on for `clang-doc`, just make it consistent.

 I'd also reword the section a bit more, and include more detail. Clang-doc is parsing each source file/TU into an AST via libtooling, then visiting important nodes in the AST to collect type and source location information for the various declarations. It's also visiting comment nodes in the AST an collecting all of those in to the various *Info bits you worked with. From there the serialized bitcode representation of the documentation Info is collected via a Map-Reduce algorithm to generate the final documentation. This process -- which you should be familiar with -- should be explained here.